source/extensions/tracers/zipkin/span_context_extractor.cc

```diff
@@ -0,0 +1,214 @@
+#include "extensions/tracers/zipkin/span_context_extractor.h"
+
+#include "common/common/assert.h"
+#include "common/common/utility.h"
+
+#include "extensions/tracers/zipkin/span_context.h"
+#include "extensions/tracers/zipkin/zipkin_core_constants.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace Zipkin {
+namespace {
+constexpr int FormatMaxLength = 32 + 1 + 16 + 3 + 16; // traceid128-spanid-1-parentid
+bool validSamplingFlags(char c) {
+  if (c == '1' || c == '0' || c == 'd')
+    return true;
+  return false;
+}
+
+bool getSamplingFlags(char c, const Tracing::Decision tracing_decision) {
+  if (validSamplingFlags(c)) {
+    return c == '0' ? false : true;
+  } else {
+    return tracing_decision.traced;
+  }
+}
+
+} // namespace
+
+SpanContextExtractor::SpanContextExtractor(Http::HeaderMap& request_headers)
+    : request_headers_(request_headers) {}
+
+SpanContextExtractor::~SpanContextExtractor() {}
+
+bool SpanContextExtractor::extractSampled(const Tracing::Decision tracing_decision) {
+  bool sampled(false);
+  if (request_headers_.B3()) {
+    absl::string_view b3 = request_headers_.B3()->value().getStringView();
+    if (b3.length() == 1) { // possibly sampling falgs
+      return getSamplingFlags(b3[0], tracing_decision);
+    }
+
+    if (b3.length() >= (32 + 1 + 16 + 2) && b3[32] == '-' && b3[49] == '-') {
+      return getSamplingFlags(b3[50], tracing_decision);
+    }
+
+    if (b3.length() >= (16 + 1 + 16 + 2) && b3[16] == '-' && b3[33] == '-') {
```

Comment:
it is also possible that there is length `16 + 1 + 16` or `32 + 1 + 16` which is a sampling decision has not yet been made.

Reply:
fixed